### PR TITLE
contribution / Add predicates to conditions converter

### PR DESCRIPTION
**Pull request by infehm**
# Description:

Adding a helper module in utils to convert a kartothek read dataset predicate to a query cube predicate.
This functionalities makes integrations from querying datasets to querying cubes easier, and also will be handy for libraries loading both data from datasets and cubes.

- [ ] Closes #xxxx
- [ ] Changelog entry

### CODE REVIEW

File: kartothek/utils/predicate_converter.py
Comment (xhochy) [inline kartothek/utils/predicate_converter.py:5]: Kartothek is Python 3+ only, so this should not be needed.
```suggestion
```

File: kartothek/utils/predicate_converter.py
Comment (xhochy) [inline kartothek/utils/predicate_converter.py:33]: Could you use `pa.types.is_int*` for this and the following?

**Review by xhochy — CHANGES_REQUESTED** [review]

File: kartothek/utils/predicate_converter.py
Comment (infehm) [reply kartothek/utils/predicate_converter.py:33]: would that make sense in this context? It should not check the type, but rather define it for the the "to_string" conversion.  
If converting to pa.types.is_int(t), what would I need to pass here as "t"?

File: kartothek/utils/predicate_converter.py
Comment (infehm) [reply kartothek/utils/predicate_converter.py]: changed

File: kartothek/utils/predicate_converter.py
Comment (xhochy) [reply kartothek/utils/predicate_converter.py:31]: Ah, sorry, I saw a `==` here. Than this is OK but I would be curious why you are defaulting to `int16` and not `int64` here?

Comment (infehm) [reply kartothek/utils/predicate_converter.py:31]: no particular reason for that, I saw this was used in some modules (but 32 and 64 as well). I will change it to int64

**Comment by xhochy** [comment]
Looks OK from my side now but my cube knowledge is limited thus waiting for someone else to pick up review from here. Probably this would also be better placed in `kartothek.core.cube.utils` instead of the global `utils` module.

File: kartothek/utils/predicate_converter.py
Comment (stephan-hesselmann-by) [inline kartothek/utils/predicate_converter.py:11]: Can you add typehints?

File: kartothek/utils/predicate_converter.py
Comment (xhochy) [inline kartothek/utils/predicate_converter.py:13]:
```suggestion
def write_predicate_as_cube_condition(predicate: Tuple[str, str, Any]) -> Condition:
```

File: kartothek/utils/predicate_converter.py
Comment (xhochy) [inline kartothek/utils/predicate_converter.py:63]:
```suggestion
def convert_predicates_to_cube_conditions(predicates: Sequence[Tuple[str, str, Any]]) -> Sequence[Condition]:
```

File: kartothek/utils/predicate_converter.py
Comment (xhochy) [inline kartothek/utils/predicate_converter.py:80]: Probably simpler as
```suggestion
    return [write_predicate_as_cube_condition(predicate) for predicate in predicates]
```

Comment (infehm) [reply kartothek/utils/predicate_converter.py:82]: This does not work as it looses functionality, but I found a working solution now for the condition type

File: kartothek/utils/predicate_converter.py
Comment (xhochy) [inline kartothek/utils/predicate_converter.py:64]: The interesting part here is the typing of the inner list. Reading the docstring I'm confused whether the inner but should be a list as it does contain exactly one element and thus should rather be this one element directly.

File: kartothek/utils/predicate_converter.py
Comment (stephan-hesselmann-by) [inline kartothek/utils/predicate_converter.py:80]: Predicates are defined as [[cond1 AND cond2] OR [cond3 AND cond4]]. They are implemented as a List of Lists, where the inner list denotes AND and the outer OR conjunctions.`predicates[0]` would therefore only take the first AND condition block and convert it. Is that intentional? I don't think that is clear from the function definition or docstring. I think you should either implement this by taking only a `List` of AND conditions in the first place (instead of `List[List[...]`), or implement the conversion of all predicates.

File: kartothek/utils/predicate_converter.py
Comment (stephan-hesselmann-by) [inline kartothek/utils/predicate_converter.py:30]: I would suggest to use Python f-strings here to be more in line with the rest of the code base

File: tests/utils/test_predicate_converter.py
Comment (stephan-hesselmann-by) [inline tests/utils/test_predicate_converter.py:11]: You can use parametrized tests here to reduce the asserts per test. I would also suggest to separate the `test_raises_when...` into separate tests.

File: tests/utils/test_predicate_converter.py
Comment (stephan-hesselmann-by) [inline tests/utils/test_predicate_converter.py:50]: Same here

**Review by stephan-hesselmann-by — APPROVED** [review]
Thanks Inga looks good now. Will merge when the master has been sorted out

File: kartothek/utils/predicate_converter.py
Comment (stephan-hesselmann-by) [inline kartothek/utils/predicate_converter.py:27]:
```suggestion
    condition: Condition
```

Comment (stephan-hesselmann-by) [reply kartothek/utils/predicate_converter.py:27]: @infehm this is ready to be merged now. Can you add a changelog entry for this change? I think the failing doc build can be fixed the suggestion

Comment (infehm) [reply kartothek/utils/predicate_converter.py]: I guess I also need to change it for the below function, right?  (convert_predicates_to_cube_conditions)

**Pull request merged**
